master/buildbot/reporters/bitbucket.py

```diff
@@ -0,0 +1,93 @@
+# This file is part of Buildbot.  Buildbot is free software: you can
+# redistribute it and/or modify it under the terms of the GNU General Public
+# License as published by the Free Software Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+import json
+
+from twisted.internet import defer
+from twisted.python import log
+
+from buildbot.process.results import SUCCESS
+from buildbot.reporters import http
+
+# Magic words understood by Butbucket REST API
+BITBUCKET_INPROGRESS = 'INPROGRESS'
+BITBUCKET_SUCCESSFUL = 'SUCCESSFUL'
+BITBUCKET_FAILED = 'FAILED'
+
+
+class BitbucketStatusPush(http.HttpStatusPushBase):
+    name = "BitbucketStatusPush"
+
+    @defer.inlineCallbacks
+    def reconfigService(self, oauth_key, oauth_secret, base_url='https://api.bitbucket.org/2.0/repositories',
+                        oauth_url='https://bitbucket.org/site/oauth2/access_token', **kwargs):
+        yield http.HttpStatusPushBase.reconfigService(self, **kwargs)
+        if base_url.endswith('/'):
+            base_url = base_url[:-1]
+        self.base_url = base_url
+        self.oauth_url = oauth_url
+        self.auth = (oauth_key, oauth_secret)
+
+    @defer.inlineCallbacks
+    def send(self, build):
+        results = build['results']
+        if build['complete']:
+            status = BITBUCKET_SUCCESSFUL if results == SUCCESS else BITBUCKET_FAILED
+        else:
+            status = BITBUCKET_INPROGRESS
+        for sourcestamp in build['buildset']['sourcestamps']:
+            sha = sourcestamp['revision']
+            body = {'state': status,
+                    'key': build['builder']['name'],
+                    'name': build['builder']['name'],
+                    'url': build['url']
+                    }
+            owner, repo = self.get_owner_and_repo(sourcestamp['repository'])
+            token = ''
+            oauth_request = yield self.session.post(self.oauth_url, auth=self.auth,
```

Comment:
I think you should get the credentials outside of the loop and remember them. Usually the token has a expiration date, so another way is to cache the token, check for validity in the beggining of the loop, and only get a new token when really necessary